src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -254,3 +255,505 @@ g.test('inputs')
+      layout: pipeline.getBindGroupLayout(0),
+      entries: [
+        {
+          binding: 0,
+          resource: {
+            buffer: sizesBuffer,
+          },
+        },
+        {
+          binding: 1,
+          resource: {
+            buffer: compareBuffer,
+          },
+        },
+      ],
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bg);
+    pass.dispatchWorkgroups(t.params.numWGs, 1, 1);
+    pass.end();
+    t.queue.submit([encoder.finish()]);
+
+    const sizesReadback = await t.readGPUBufferRangeTyped(sizesBuffer, {
+      srcByteOffset: 0,
+      type: Uint32Array,
+      typedLength: numInvocations,
+      method: 'copy',
+    });
+    const sizesData: Uint32Array = sizesReadback.data;
+
+    const compareReadback = await t.readGPUBufferRangeTyped(compareBuffer, {
+      srcByteOffset: 0,
+      type: Uint32Array,
+      typedLength: numInvocations,
+      method: 'copy',
+    });
+    const compareData: Uint32Array = compareReadback.data;
+
+    t.expectOK(checkSubgroupSizeConsistency(sizesData, compareData, minSize, maxSize, wgThreads));
+  });
+
+/**
+ * Checks the consistency of subgroup_invocation_id builtin values
+ *
+ * @param data The subgroup_invocation_id output data
+ * @param
+ */
+function checkSubgroupInvocationIdConsistency(
+  data: Uint32Array,
+  ids: Uint32Array,
+  subgroupSize: number,
+  invocations: number,
+  numWGs: number
+): Error | undefined {
+  for (let wg = 0; wg < numWGs; wg++) {
+    const mappings = new Map();
+    for (let i = 0; i < invocations; i++) {
+      const idx = i + invocations * wg;
+      const subgroup_id = ids[idx];
+      if (subgroup_id === 999) {
+        return new Error(`Invocation ${i}: no data`);
+      }
+      if (!mappings.has(subgroup_id)) {
+        mappings.set(subgroup_id, 0n);
+      }
+      let v = mappings.get(subgroup_id);
+      v |= 1n << BigInt(data[idx]);
+      mappings.set(subgroup_id, v);
+    }
+
+    for (const value of mappings.entries()) {
+      const id = value[0];
+      const ballot = value[1];
+
+      let onebits = popcount(Number(BigInt.asUintN(32, ballot)));
+      onebits += popcount(Number(BigInt.asUintN(32, ballot >> 32n)));
+      onebits += popcount(Number(BigInt.asUintN(32, ballot >> 64n)));
+      onebits += popcount(Number(BigInt.asUintN(32, ballot >> 96n)));
+      if (onebits > subgroupSize) {
+        return new Error(
+          `Subgroup including invocation ${id} is too large, ${onebits}, for subgroup size, ${subgroupSize}`
+        );
+      }
+
+      const ballotP1 = ballot + 1n;
+      onebits = popcount(Number(BigInt.asUintN(32, ballotP1)));
+      onebits += popcount(Number(BigInt.asUintN(32, ballotP1 >> 32n)));
+      onebits += popcount(Number(BigInt.asUintN(32, ballotP1 >> 64n)));
+      onebits += popcount(Number(BigInt.asUintN(32, ballotP1 >> 96n)));
+      if (onebits !== 1) {
```

Comment:
I understand this is validating `ballot` being `2**K - 1` for some `K`.
If subgroup size is 128 (although we have never seen this case in field), I thought `ballotP1` would be `2**128` and `onebits` become `0` instead of `1`, since we only add up the popcount of lower 128 bits?

Maybe we can use
``` typescript
const ballotP1 = ballot + 1n;
if ((ballotP1 & ballot) !== 0n) {
  ...
}
```
Since a positive integer `x` is a power of 2 if and only if `(x & (x-1)) === 0`.